File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/build_execution_deployment/BuildExecutionDeploymentSettingsApplier.java
package com.github.alexandrecarlton.idea.settings.applier.impl.build_execution_deployment;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.BuildExecutionDeploymentSettings;
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.build_tools.BuildToolsSettings;
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.compiler.CompilerSettings;
import com.intellij.externalDependencies.DependencyOnPlugin;
import com.intellij.externalDependencies.ExternalDependenciesManager;

import javax.inject.Inject;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

public class BuildExecutionDeploymentSettingsApplier implements SettingsApplier<BuildExecutionDeploymentSettings> {

  private final ExternalDependenciesManager externalDependenciesManager;
  private final SettingsApplier<BuildToolsSettings> buildToolsSettingsApplier;
  private final SettingsApplier<CompilerSettings> compilerSettingsApplier;

  @Inject
  public BuildExecutionDeploymentSettingsApplier(ExternalDependenciesManager externalDependenciesManager,
                                                 SettingsApplier<BuildToolsSettings> buildToolsSettingsApplier,
                                                 SettingsApplier<CompilerSettings> compilerSettingsApplier) {
    this.externalDependenciesManager = externalDependenciesManager;
    this.buildToolsSettingsApplier = buildToolsSettingsApplier;
    this.compilerSettingsApplier = compilerSettingsApplier;
  }

  @Override
  public void apply(BuildExecutionDeploymentSettings settings) {
    externalDependenciesManager.setAllDependencies(Stream.concat(
        externalDependenciesManager.getAllDependencies().stream(),
        settings.requiredPlugins()
            .stream()
            .map(required -> new DependencyOnPlugin(
                required.plugin(),
                required.minimumVersion().orElse(null),
                required.maximumVersion().orElse(null))))
        .collect(toList()));

    settings.buildTools().ifPresent(buildToolsSettingsApplier::apply);
    settings.compiler().ifPresent(compilerSettingsApplier::apply);
  }
}

File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/build_execution_deployment/BuildExecutionDeploymentSettingsApplier.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.build_execution_deployment

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.BuildExecutionDeploymentSettings
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.build_tools.BuildToolsSettings
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.compiler.CompilerSettings
import com.intellij.externalDependencies.DependencyOnPlugin
import com.intellij.externalDependencies.ExternalDependenciesManager
import javax.inject.Inject

class BuildExecutionDeploymentSettingsApplier @Inject
constructor(
    private val externalDependenciesManager: ExternalDependenciesManager,
    private val buildToolsSettingsApplier: SettingsApplier<BuildToolsSettings>,
    private val compilerSettingsApplier: SettingsApplier<CompilerSettings>
) : SettingsApplier<BuildExecutionDeploymentSettings> {

    override fun apply(settings: BuildExecutionDeploymentSettings) {
        externalDependenciesManager.allDependencies = settings.requiredPlugins()
            .map { required ->
                DependencyOnPlugin(
                    required.plugin(),
                    required.minimumVersion().orElse(null),
                    required.maximumVersion().orElse(null))
            }

        settings.buildTools().ifPresent { buildToolsSettingsApplier.apply(it) }
        settings.compiler().ifPresent { compilerSettingsApplier.apply(it) }
    }
}
